maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/metadata/AddPluginArtifactMetadataMojo.java

```diff
@@ -69,25 +69,33 @@ public class AddPluginArtifactMetadataMojo extends AbstractMojo {
     @Parameter(defaultValue = "false", property = "maven.plugin.skip")
     private boolean skip;
 
+    @Component
+    private RuntimeInformation runtimeInformation;
+
+    private final VersionScheme versionScheme = new GenericVersionScheme();
+
     /** {@inheritDoc} */
     @Override
     public void execute() throws MojoExecutionException {
         if (skip) {
             getLog().warn("Execution skipped");
             return;
         }
-        Artifact projectArtifact = project.getArtifact();
-
-        Versioning versioning = new Versioning();
-        versioning.setLatest(projectArtifact.getVersion());
-        versioning.updateTimestamp();
-        ArtifactRepositoryMetadata metadata = new ArtifactRepositoryMetadata(projectArtifact, versioning);
-        projectArtifact.addMetadata(metadata);
-
-        GroupRepositoryMetadata groupMetadata = new GroupRepositoryMetadata(project.getGroupId());
-        groupMetadata.addPluginMapping(getGoalPrefix(), project.getArtifactId(), project.getName());
+        // nothing if Maven is 3.9+
+        try {
+            if (versionScheme
+                            .parseVersion("3.9.0")
+                            .compareTo(versionScheme.parseVersion(runtimeInformation.getMavenVersion()))
+                    < 1) {
+                getLog().info("This Mojo is not used in Maven version 3.9.0 and above");
+                return;
+            }
+        } catch (InvalidVersionSpecificationException e) {
+            // not happening with generic
+            throw new MojoExecutionException(e);
+        }
 
-        projectArtifact.addMetadata(groupMetadata);
+        LegacySupport.execute(project, getGoalPrefix());
```

Comment:
We should be careful with it .... probably it is used by Nexus Staging Plugin
https://issues.apache.org/jira/browse/MPLUGIN-384

Reply:
I consider nexus-staging-maven-plugin legacy, simply put, people should avoid using it... all the benefits it (may) add, is that you don't have to log into Nx UI and press "close" button. Simply put, not worth if you contrast that to legacy burden it brings and have.

Reply:
I fully agree with @cstamas and ignore it.

Reply:
👍

Reply:
We will see how many people use nexus-staging

Reply:
Is it materialized .... https://issues.apache.org/jira/browse/MNG-8121 ❓

Reply:
This change is unrelated. What happened is that those users had no prefix set (ever), so remote metadata has no prefix either, and now that they upgraded, and forced to set plugin prefix (in latest m-p-p), NPE happens.